### PR TITLE
Add Itai Rodeh asynchronous directed ring leader election algorithm

### PR DESCRIPTION
**Pull request by Onchao**
Function runner.Run(...) is now parametrized by boolean value, which controls closing Nodes.

### CODE REVIEW

File: leader/directed_ring/async_itai_rodeh.go [hunk 1]
```diff
@@ -45,128 +48,119 @@ func getStateAsyncItaiRodeh(node lib.Node) stateAsyncItaiRodeh {
...
 
-	runner.Run()
-	//checkSingleLeaderElected(vertices) // TODO
-
+	runner.Run(false)
+	log.Println("Single leader selected: ", checkSingleLeaderElected(vertices))
 	return runner.GetStats()
 }
+
+func wtf() {
```
Comment (krzysztof-turowski) [inline leader/directed_ring/async_itai_rodeh.go:164]: This should be removed

File: text/1126923/async_itai_rodeh.tex [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+\begin{itemize}
+    \item $id$ --- information about the original sender
+    \item $round$ --- information about the original sender
+    \item $hop$ --- distance that a message travelled
+    \item $bit$ --- whether there exist identical process to the original sender (need to select new identity)
+    \item[$\star$] additionally message id of the preceding process is send to properly terminate
+\end{itemize}
+
+\subsubsection*{Procedures}
+Precise description can be found on page 5 here:\cite{fokkink2005simplifying}. I will only describe an algorithm with words for simplicity. \\
```
Comment (krzysztof-turowski) [inline text/1126923/async_itai_rodeh.tex:63]: on page 5 in \cite{...} + please remove usages of \\

File: text/1126923/async_itai_rodeh.tex [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    \item else purges the message
+\end{itemize}
+
+\clearpage
+
+\subsection*{Correctness}
+First, let's consider a simpler case in which the algorithm was lucky, and each process's identity is different. Then we would get a problem of leader election for a unidirectional ring with asynchronous communication but this time with unique identities. No $bit$ would change. No process would advance to round 2. Itai-Rodeh algorithm would be reduced to Chang and Roberts algorithm for leader selection.
+
+In the general case, an algorithm has $round$'s and $bit$'s are tracking information about duplicate $id$'s.  On the ring, until a leader is elected there always exist a message with lexicographically largest pair $(round,\ id)$ that travels to its owner (maybe a few of them). If there exists at least one process lexicographically smaller than this message and it reaches him, then one more process will become passive bringing us a step closer to electing a leader. Otherwise, all the processes will move to the next round and with some luck, some processes will become passive. Eventually, the leader will be elected.
```
Comment (krzysztof-turowski) [inline text/1126923/async_itai_rodeh.tex:83]: exist a message -> exists a message

File: text/1126923/async_itai_rodeh.tex [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+\usepackage{cleveref}
+\newtheorem{lemma}{Lemma}
+\newtheorem{theorem}{Theorem}
+\begin{document}
+\section*{Itai-Rodeh Leader Election for Anonymous Rings}
+
+\subsection*{Introduction}
+Itai-Rodeh leader election algorithm is a probabilistic leader election algorithm for anonymous unidirectional rings with asynchronous communication. Additionally, the size of the ring is known to all processes.
+
+Processes select random identities from a finite domain (of size bigger than $1$). Processes with the largest identities start another new election rounds, select new identities, and continue broadcasting messages until a leader is elected. Processes with smaller identities became passive that only pass on the messages they receive. It is assumed that the size of the ring is known to all processes so that each process can recognize its own message. The Itai-Rodeh algorithm terminates with probability one, and eventually, exactly one leader is elected. Expected number of messages is at most $O(n^2)$.
```
Comment (krzysztof-turowski) [inline text/1126923/async_itai_rodeh.tex:42]: Expected number -> The expected number

**Review by krzysztof-turowski — CHANGES_REQUESTED** [review]
Thank you very much, the PR is almost ready.

Two additional things to do:
1. change the name of the pull request,
2. add your algorithm to README.md

File: leader/directed_ring/async_itai_rodeh.go (hunk 1 above)
Comment (Onchao) [reply leader/directed_ring/async_itai_rodeh.go:164]: :D

**Review by krzysztof-turowski — CHANGES_REQUESTED** [review]
Please change the name of PR for something more resembling previous PRs and please update the description of PR, it still says "work in progress" (which it isn't as for now), but nothing about its content.

**Review by krzysztof-turowski — APPROVED** [review]

**Pull request merged**